draft-ietf-quic-transport.md

```diff
@@ -2352,171 +2361,136 @@ shown in the following figure and described below.
+received.
 
-A stream is "half-closed (remote)" when the stream is no longer being used by
-the peer to send any data.  An endpoint will have either received all data that
-a peer has sent or will have received a RST_STREAM frame and discarded any
-received data.
 
-Once all data has been either received or discarded, a sender is no longer
-obligated to update the maximum received data for the connection.
+### The "closed" State
 
-An endpoint that receives a RST_STREAM frame (and which has not sent a FIN or a
-RST_STREAM) MUST immediately respond with a RST_STREAM frame, and MUST NOT send
-any more data on the stream.
+A "closed" stream cannot be used for sending any stream-related frames.
 
-Due to reordering, an endpoint could continue receiving frames for the stream
-even after the stream is closed for sending.  Frames received after a peer
-closes a stream SHOULD be discarded.  An endpoint MAY choose to limit the period
-over which it ignores frames and treat frames that arrive after this time as
-being in error.
+In the "closed" state, a sending endpoint can retransmit data that has already
+been sent in STREAM frames, or retransmit a RST_STREAM frame.  Retransmission of
+data is permitted if the packets carrying the original data are determined to
+have been lost.
 
 An endpoint will know the final offset of the data it receives on a stream when
-it reaches the "half-closed (remote)" state, see {{final-offset}} for details.
-
-A stream in this state can be used by the endpoint to send any frame that
-mentions a stream ID.  In this state, the endpoint MUST observe advertised
-stream and connection data limits (see {{flow-control}}).
+it reaches the "closed" state, see {{final-offset}} for details.
 
-A stream transitions from this state to "closed" by completing transmission of
-all data.  This includes sending all data carried in STREAM frames up including
-the terminal STREAM frame that contains a FIN flag.
-
-A stream becomes "closed" when the endpoint sends and receives acknowledgment of
-a RST_STREAM frame.
-
-
-### closed {#state-closed}
-
-The "closed" state is the terminal state for a stream.
-
-Once a stream reaches this state, no frames can be sent that mention the stream.
-Reordering might cause frames to be received after closing, see
-{{state-hc-remote}}.
+An endpoint could continue receiving frames for the stream even after the stream
+is closed for sending if packets are reordered.  At the receiving endpoint,
+retransmissions of data in STREAM frames, or retransmissions of STREAM_BLOCKED
+and RST_STREAM frames might arrive after the stream becomes "closed".  At the
+sending endpoint, MAX_STREAM_DATA frames might be received after the stream
+becomes "closed".  Frames received on a "closed" stream SHOULD be discarded.  An
+endpoint MAY choose to limit the period over which it ignores frames and treat
+frames that arrive after this time as being in error.
 
```

Comment:
Perhaps a note on fatal connection error if receiving frames at later offset than close - otherwise a transmitter might abuse flow control limits - though the benefit of that is unclear. Perhaps also a note here stating the both RST_STREAM and FIN defines a specific end offset - even if discussed in more detail in specific frames.